# Исключения
**обработка исключений:**
- `try`
- `catch`
- `throw`
- `throws`
- `finally`

стандартный случай:
```
try {
    ....
} catch {
    обработка
} finally {
    фрагмент кода, который исполняется 
    ВООБЩЕ всегда для 99.9999999% случаев. 
    есть маленький шанс встретить System.exit(); 
    вы прибили JVM кувалдой
}
```
catch с параметром:

```
try {
    ...
} catch (_че-то throwable_) {
    обработка
}
```

**throwable:**
- error (все плохо очень)
- exception (можно что-то исправить)
    - `RuntimeException` ((и его наследники) проверяемый
    - other...


**исключения:**
- проверяемые (компилятор проверяет)
- не проверяемые (на нашей совести)

**способы обработки:**
1. try-catch
2. throws (если хотим скинуть обработку на уровень выше)

   `void foo() throws IOException`


throw - так же, как в плюсах:
```
try {
    ...
} catch {
    throw new IOException();
}
```
можно ловить разные типы исключений:
```
try {
    ...
} catch (IOException e) {
    ...
} catch (_another_exception_type_ e2) { ... }
```
> **(!)** писать `catch (Exception e)` для нескольких кетчей не надо, т.к. `Exception` сам по себе хватает любые исключения, код с конкретными исключениями просто не будет достигнут

мавен градл [не помню, в каком контексте :D]

# чтение из файлов:

2 больших библиотеки
- io
- nio (new io :D)

в контексте исключений:
```
try {
    FileReader r = // конструктор
    // читаем
    // r.close(); - тоже кидает исключения
} catch (IOExceptoin e) {
    ...
} finally {
    try {
        r.close();
    } catch (...) {
        ...
        // появится еще куча вложенных кетчей
    }
}
```

**как нормально?**

`try (FileReader r = .....) { // try с ресурсами`

> (!) аргумент должен реализовывать интерфейс AutoClosable

способ фармить -баллы:
`psvm(Str[]) throws Ex`